### DISCUSSION TITLE
Announcing gptstudio v0.4.0: Enhanced AI Integration in Your RStudio IDE

### DISCUSSION THREAD

**Discussion by calderonsamuel** [MichelNivard/gptstudio#210] [Show and tell]
We are excited to announce the release of `{gptstudio}` version 0.4.0, which brings a host of new features and improvements designed to enrich your AI-assisted coding experience in RStudio. Building on the robust foundation of our previous releases, this update enhances the functionality and user experience of the Chat app while introducing support for additional AI models and services.

Try it out with:

```r
install.packages("gptstudio")
```

After that, go to `Addins > GPTSTUDIO > Chat` or open the Rstudio command palette (Ctrl + Shift + P on Windows) and search for it.

The Chat app has always been a powerful tool for engaging with AI models directly within RStudio. It offers an interactive Shiny application that enables you to converse with various AI models, providing a seamless interface to type questions, code snippets, or text and receive AI-generated responses. This makes it an invaluable asset for querying coding advice, explanations of complex concepts, and generating code snippets. 

## Layout change

In version 0.4.0, we introduce several significant enhancements to the Chat app. The most notable is a more minimalistic starting view. To provide as mush space as posible for your chat interactions.

![image](https://github.com/MichelNivard/gptstudio/assets/19418298/049ef7a4-83a3-4f04-8c71-0d1aab6d6411)

This comes with the addition of a sidebar, powered by `{bslib}`, which allows users to easily access their conversation history, start new chats, and adjust settings. 

![image](https://github.com/MichelNivard/gptstudio/assets/19418298/1ed24e2a-d322-41a4-bb81-c47acaeaaa94)

This redesign provides more space for displaying messages and includes tooltips to assist with navigation. Chats are now automatically saved and updated after each AI response, and users can edit the titles of their conversations for better organization. This update makes the Chat app more user-friendly and efficient, enhancing your overall experience.

## Support for local models

We are also thrilled to announce the support for local models through Ollama. This allows users to run AI models locally, providing greater control and customization. The default host is set to http://localhost:11434, but it can be adjusted via the OLLAMA_HOST environmental variable. Users are responsible for maintaining their own Ollama installations and models, giving them the flexibility to manage their local AI resources.

You can run ollama in any platform as a docker container. The following code runs the CPU-only version:

```bash
docker run -d -v ollama:/root/.ollama -p 11434:11434 --name ollama ollama/ollama
```

You can see more docker options in the Ollama [official blog post](https://ollama.ai/blog/ollama-is-now-available-as-an-official-docker-image).

Before using the service, you need to pull a model. Run the following code inside your container to pull llama2:

```bash
ollama pull llama2
```

Now you can use it with chat addin going to `Settings > API service > Select API Service > ollama`.

Check the [ollama library](https://ollama.ai/library) to see more models. For more advanced install options, check the [official documentation](https://github.com/jmorganca/ollama).

## Reading R help pages

The chat addin can now provide the help page for R functions or objects when the user message contains the construct `package::object`. This is useful for getting help on the latest changes in the R ecosystem, specially those posterior to the cut-off date of the different LLMs. 

Look what the chat assistant would normally respond when asked about `dplyr::join_by()`:

<img src="https://github.com/MichelNivard/gptstudio/assets/19418298/6844f644-a51f-4c73-adee-cd6c6056a363" width="600px">

Now compare the response when this new option is activated:

<img src="https://github.com/MichelNivard/gptstudio/assets/19418298/84841f97-c5a4-4374-a321-1b6aac443cce" width="600px">

The "R documentation" message was automatically added by the chat app and it passes the help page of the function to the service's API behind the scenes. To activate it, you need to go `Settings > Assistant behavior > Read help pages` and save as default.  Beware that this will increase the token count of your API requests and that some R packages document their exported objects with different naming conventions.

## Support for even more services

Moreover, gptstudio now integrates with Perplexity and Cohere services, expanding the range of available AI models. Perplexity offers a variety of models, including llama-3-sonar and mixtral-8x7b, while Cohere provides models such as command and command-light. These integrations allow users to choose from a broader array of AI solutions, tailoring their setup to best meet their specific needs.

To further improve usability, we have introduced a new function, `gpstudio_sitrep()`, which helps with debugging and setup. Additionally, new vignettes have been added to guide users through the setup of each service, ensuring comprehensive documentation and support.

| AI Service                                                                                           | Models                                                                                                                                                 | Documentation                                                                                                                           | Setup                                                                                   |
|------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
| [OpenAI](https://platform.openai.com)                                                                | gpt-4-turbo, gpt-4, gpt-3.5-turbo ([latest models](https://platform.openai.com/docs/models))                                                           | [OpenAI API Docs](https://platform.openai.com/docs/api-reference)                                                                       | [OpenAI Setup](https://michelnivard.github.io/gptstudio/articles/openai.html)           |
| [HuggingFace](https://huggingface.co/)                                                               | various                                                                                                                                                | [HF Inference API Docs](https://huggingface.co/docs/hub/models-inference)                                                               | [HF Setup](https://michelnivard.github.io/gptstudio/articles/huggingface.html)          |
| [Anthropic](https://docs.anthropic.com/claude/docs/guide-to-anthropics-prompt-engineering-resources) | claude-2.1, claude-instant-1.2                                                                                                                         | [Anthropic API Docs](https://docs.anthropic.com/claude/reference/getting-started-with-the-api)                                          | [Anthropic Setup](https://michelnivard.github.io/gptstudio/articles/anthropic.html)     |
| [Ollama](https://ollama.com/)                                                                        | mistral, llama2, mixtral, phi ([latest models](https://ollama.com/library))                                                                            | [Ollama API Docs](https://github.com/ollama/ollama/blob/main/docs/api.md)                                                               | [Ollama Setup](https://michelnivard.github.io/gptstudio/articles/ollama.html)           |
| [Perplexity](https://www.perplexity.ai)                                                              | pplx-7b-chat, pplx-70b-chat, pplx-7b-online, pplx-70b-online, llama-2-70b-chat, codellama-34b-instruct, mistral-7b-instruct, and mixtral-8x7b-instruct | [Perplexity API Docs](https://docs.perplexity.ai/reference/post_chat_completions)                                                       | [Perplexity Setup](https://michelnivard.github.io/gptstudio/articles/perplexity.html)   |
| [Google AI Studio](https://ai.google.dev/tutorials/ai-studio_quickstart)                             | Gemini and Palm (legacy)                                                                                                                               | [Google AI Studio Docs](https://ai.google.dev/docs)                                                                                     | [Google AI Studio Setup](https://michelnivard.github.io/gptstudio/articles/google.html) |
| [Azure OpenAI](https://learn.microsoft.com/en-us/azure/ai-services/openai/overview)                  | gpt-4, gpt-3.5-turbo ([latest models](https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models#gpt-4-and-gpt-4-turbo-preview))       | [Azure OpenAI API Docs](https://learn.microsoft.com/en-us/azure/ai-services/openai/quickstart?tabs=command-line,python&pivots=rest-api) | [Azure OpenAI Setup](https://michelnivard.github.io/gptstudio/articles/azure.html)      |
| [Cohere](https://cohere.com/)                                                                        | command, command-light, command-nightly, command-light-nightly                                                                                         | [Cohere API Docs](https://docs.cohere.com/)                                                                                             | [Cohere Setup](https://michelnivard.github.io/gptstudio/articles/cohere.html)           |

## Constant maintainability

Internally, this release includes several improvements and bug fixes. We have reverted to using an R6 class for OpenAI streaming to enhance stability and avoid server errors. Various bugs related to OpenAI model retrieval, Azure OpenAI request formation, and Unix platform connectivity have been addressed. The codebase now utilizes `{lintr}` more extensively for maintaining consistency, and we have made several UI and performance tweaks, including improved scrollbar design and enhanced model integration.

With gptstudio v0.4.0, we continue to enhance the integration of AI capabilities within RStudio, making it easier and more efficient for data scientists and analysts to leverage advanced models in their workflows. We encourage you to explore the new features and improvements, and we look forward to your feedback. Stay tuned for more updates and happy coding!

**Comment by andrescardonas7** [comment]
Its free
